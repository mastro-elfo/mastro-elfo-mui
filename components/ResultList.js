"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;

var _react = _interopRequireDefault(require("react"));

var _core = require("@material-ui/core");

var _ = require("./");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Organizes a List with optional SubHeader
 *
 * If `results` is `null` or `undefined`, then prints nothing.
 *
 * Otherwise `results` must be an array, for each item `mapper` is called and must provide at least a unique `key`.
 *
 * Also `LeftIcon` and `RightIcon` are used inside a `ListItemIcon`, and `primary` and `secondary` are used with `ListItemText`.
 *
 * Anything else is passed to `ListItem`. By default `onClick` sets `button` to `true`.
 * @param  {[type]} subheader [description]
 * @param  {[type]} [mapper=r => r] [description]
 * @param  {[type]} results   [description]
 * @param  {[type]} others    [description]
 * @return {[type]}           [description]
 */
function _default(_ref) {
  var subheader = _ref.subheader,
      _ref$mapper = _ref.mapper,
      mapper = _ref$mapper === void 0 ? function (r) {
    return r;
  } : _ref$mapper,
      results = _ref.results,
      others = _objectWithoutProperties(_ref, ["subheader", "mapper", "results"]);

  return /*#__PURE__*/_react["default"].createElement(_.Condition, {
    show: results !== null && results !== undefined
  }, /*#__PURE__*/_react["default"].createElement(_core.List, _extends({
    subheader: !!subheader && /*#__PURE__*/_react["default"].createElement(_core.ListSubheader, null, subheader)
  }, others), results && results.map(function (r) {
    var _mapper = mapper(r),
        LeftIcon = _mapper.LeftIcon,
        RightIcon = _mapper.RightIcon,
        primary = _mapper.primary,
        secondary = _mapper.secondary,
        onClick = _mapper.onClick,
        others = _objectWithoutProperties(_mapper, ["LeftIcon", "RightIcon", "primary", "secondary", "onClick"]);

    return /*#__PURE__*/_react["default"].createElement(_core.ListItem, _extends({
      button: !!onClick,
      onClick: onClick
    }, others), !!LeftIcon && /*#__PURE__*/_react["default"].createElement(_core.ListItemIcon, null, LeftIcon), /*#__PURE__*/_react["default"].createElement(_core.ListItemText, {
      primary: primary,
      secondary: secondary
    }), !!RightIcon && /*#__PURE__*/_react["default"].createElement(_core.ListItemIcon, null, RightIcon));
  })));
}