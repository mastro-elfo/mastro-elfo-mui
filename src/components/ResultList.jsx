import React from "react";
import PropTypes from "prop-types";

import {
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  ListSubheader
} from "@material-ui/core";

import { Condition } from "./";

/**
 * Organizes a List with optional SubHeader
 *
 * If `results` is `null` or `undefined`, then prints nothing.
 *
 * Otherwise `results` must be an array, for each item `mapper` is called and must provide at least a unique `key`.
 *
 * Also `LeftIcon` and `RightIcon` are used inside a `ListItemIcon`, and `primary` and `secondary` are used with `ListItemText`.
 *
 * Anything else is passed to `ListItem`. By default `onClick` sets `button` to `true`.
 * @param  {[type]} subheader [description]
 * @param  {[type]} [mapper=r => r] [description]
 * @param  {[type]} results   [description]
 * @param  {[type]} others    [description]
 * @return {[type]}           [description]
 */
export default function ResultList({
  mapper = r => r,
  results,
  subheader,
  ...others
}) {
  return (
    <Condition show={results !== null && results !== undefined}>
      <List
        subheader={!!subheader && <ListSubheader>{subheader}</ListSubheader>}
        {...others}
      >
        {results &&
          results.map(r => {
            const {
              LeftIcon,
              RightIcon,
              rightAction,
              primary,
              secondary,
              onClick,
              ...others
            } = mapper(r);
            // TODO: // Deprecated: RightIcon, remove in version 2
            if (!!RightIcon)
              console.warn(
                "RightIcon is deprecated and will be removed in next major version"
              );

            return (
              <ListItem button={!!onClick} onClick={onClick} {...others}>
                {!!LeftIcon && <ListItemIcon>{LeftIcon}</ListItemIcon>}

                <ListItemText primary={primary} secondary={secondary} />

                {!!rightAction && rightAction}
                {!rightAction && !!RightIcon && (
                  <ListItemIcon>{RightIcon}</ListItemIcon>
                )}
              </ListItem>
            );
          })}
      </List>
    </Condition>
  );
}

ResultList.propTypes = {
  mapper: PropTypes.func,
  results: PropTypes.array,
  subheader: PropTypes.string
};
