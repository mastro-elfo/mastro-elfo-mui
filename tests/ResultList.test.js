import React from "react";
import renderer from "react-test-renderer";

import ResultList from "../src/components/ResultList";
import { IconButton } from "@material-ui/core";
import BrokenImageIcon from "@material-ui/icons/BrokenImage";

test("Render", () => {
  const component = renderer.create(
    <ResultList
      subheader="subheader"
      results={[
        {
          primary: "Primary",
          secondary: "Secondary",
          onClick: () => {},
          LeftIcon: <BrokenImageIcon />,
          rightAction: (
            <IconButton onClick={() => {}}>
              <span />
            </IconButton>
          )
        }
      ].map((item, i) => ({ key: i, ...item }))}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
