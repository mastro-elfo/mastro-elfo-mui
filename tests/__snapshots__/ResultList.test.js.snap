// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1`] = `
<ul
  className="MuiList-root MuiList-padding MuiList-subheader"
>
  <li
    className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
  >
    subheader
  </li>
  <div
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemIcon-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M21 5v6.59l-3-3.01-4 4.01-4-4-4 4-3-3.01V5c0-1.1.9-2 2-2h14c1.1 0 2 .9 2 2zm-3 6.42l3 3.01V19c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2v-6.58l3 2.99 4-4 4 4 4-3.99z"
        />
      </svg>
    </div>
    <div
      className="MuiListItemText-root MuiListItemText-multiline"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Primary
      </span>
      <p
        className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        Secondary
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <span />
      </span>
    </button>
  </div>
</ul>
`;
